### PR TITLE
[docs][serve] add doc for keep alive timeout

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

Add docs related to `keep_alive_timeout_s`

## Related issue number

Closes: https://github.com/anyscale/product/issues/23527

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/serve/advanced-guides/performance.md [hunk 1]
```diff
@@ -103,6 +103,7 @@ There are handful of ways to address these issues:
 2. Try batching your requests. See [Dynamic Request Batching](serve-performance-batching-requests).
 3. Consider using `async` methods in your callable. See [the section below](serve-performance-async-methods).
 4. Set an end-to-end timeout for your HTTP requests. See [the section below](serve-performance-e2e-timeout).
+5. Set keep alive timeout for uvicorn server on Serve's HTTP proxy. See [the section below](serve-performance-keep-alive-timeout).
```
Comment (shrekris-anyscale) [inline doc/source/serve/advanced-guides/performance.md:106]: Is there a better place to link this? I don't think the keep alive timeout is related to performance issues. Maybe in [the HTTP guide](https://docs.ray.io/en/latest/serve/http-guide.html)?

File: doc/source/serve/advanced-guides/performance.md [hunk 2]
```diff
@@ -124,3 +125,16 @@ proper backpressure. You can increase the value in the deployment decorator; e.g
 By default, Serve lets client HTTP requests run to completion no matter how long they take. However, slow requests could bottleneck the replica processing, blocking other requests that are waiting. It's recommended that you set an end-to-end timeout, so slow requests can be terminated and retried.
 
 You can set an end-to-end timeout for HTTP requests by setting the `request_timeout_s` in the `http_options` field of the Serve config. HTTP Proxies will wait for that many seconds before terminating an HTTP request. This config is global to your Ray cluster, and it cannot be updated during runtime. Use [client-side retries](serve-best-practices-http-requests) to retry requests that time out due to transient failures.
+
+(serve-performance-keep-alive-timeout)=
+### Set keep alive timeout
+
+Serve uses Uvicorn HTTP server internally to serve HTTP requests. By default, Serve sets
```
Comment (shrekris-anyscale) [inline doc/source/serve/advanced-guides/performance.md:132]:
```suggestion
Serve uses a Uvicorn HTTP server internally to serve HTTP requests. By default, Serve sets
```

File: doc/source/serve/advanced-guides/performance.md [hunk 3]
```diff
@@ -124,3 +125,16 @@ proper backpressure. You can increase the value in the deployment decorator; e.g
...
+Uvicorn keeps HTTP connections alive for 5 seconds. The setting defines the duration to
+keep idle connections alive when no requests are ongoing. You can set the keep alive
+timeout by setting the `keep_alive_timeout_s` in the `http_options` field of the Serve
```
Comment (shrekris-anyscale) [inline doc/source/serve/advanced-guides/performance.md:135]:
```suggestion
Serve uses Uvicorn HTTP server internally to serve HTTP requests. By default,
Uvicorn keeps HTTP connections alive for 5 seconds between requests. Modify the keep-alive
timeout by setting the `keep_alive_timeout_s` in the `http_options` field of the Serve
```

File: doc/source/serve/advanced-guides/performance.md [hunk 4]
```diff
@@ -124,3 +125,16 @@ proper backpressure. You can increase the value in the deployment decorator; e.g
...
+config files. This config is global to your Ray cluster, and it cannot be updated during
+runtime. You can also set `RAY_SERVE_HTTP_KEEP_ALIVE_TIMEOUT_S` environment variable to
+set the keep alive timeout. `RAY_SERVE_HTTP_KEEP_ALIVE_TIMEOUT_S` will take the
+precedent over the `keep_alive_timeout_s` config if both are set. You can find more
```
Comment (shrekris-anyscale) [inline doc/source/serve/advanced-guides/performance.md:139]:
```suggestion
set the keep alive timeout. `RAY_SERVE_HTTP_KEEP_ALIVE_TIMEOUT_S` will take the
precedence over the `keep_alive_timeout_s` config if both are set. You can find more
```

File: doc/source/serve/advanced-guides/performance.md (hunk 1 above)
Comment (GeneDer) [reply doc/source/serve/advanced-guides/performance.md:106]: Yea, I feel it's related to the request timeout so put it here. Will move to http guide

File: doc/source/serve/http-guide.md [hunk 5]
```diff
@@ -295,3 +296,15 @@ Here is a list of adapters; please feel free to [contribute more](https://github
...
+### Set keep alive timeout
+
+Serve uses a Uvicorn HTTP server internally to serve HTTP requests. By default, Uvicorn
+keeps HTTP connections alive for 5 seconds between requests. Modify the keep-alive
+timeout by setting the `keep_alive_timeout_s` in the `http_options` field of the Serve
+config files. This config is global to your Ray cluster, and it cannot be updated during
+runtime. You can also set `RAY_SERVE_HTTP_KEEP_ALIVE_TIMEOUT_S` environment variable to
+set the keep alive timeout. `RAY_SERVE_HTTP_KEEP_ALIVE_TIMEOUT_S` will take the
+precedence over the `keep_alive_timeout_s` config if both are set. You can find more
+information about Uvicorn's keep alive timeout [here](https://www.uvicorn.org/settings/#keep-alive-timeout).
```
Comment (zcin) [inline doc/source/serve/http-guide.md:310]: I think right now this is under `HTTP Adapters`:
image

Comment (GeneDer) [reply doc/source/serve/http-guide.md:310]: oops, thanks for catching it!

File: doc/source/serve/http-guide.md [hunk 6]
```diff
@@ -6,6 +6,7 @@ This section helps you understand how to:
 - use Ray Serve to integrate with FastAPI
```
Comment (angelinalg) [inline doc/source/serve/http-guide.md:6]:
```suggestion
- Use Ray Serve to integrate with FastAPI
```

File: doc/source/serve/http-guide.md [hunk 7]
```diff
@@ -6,6 +6,7 @@ This section helps you understand how to:
...
 - use customized HTTP adapters
```
Comment (angelinalg) [inline doc/source/serve/http-guide.md:7]:
```suggestion
- Use customized HTTP adapters
```

File: doc/source/serve/http-guide.md [hunk 8]
```diff
@@ -6,6 +6,7 @@ This section helps you understand how to:
...
 - choose which feature to use for your use case
```
Comment (angelinalg) [inline doc/source/serve/http-guide.md:8]:
```suggestion
- Choose which feature to use for your use case
```

File: doc/source/serve/http-guide.md [hunk 9]
```diff
@@ -6,6 +6,7 @@ This section helps you understand how to:
...
+- setup keep alive timeout
```
Comment (angelinalg) [inline doc/source/serve/http-guide.md:9]:
```suggestion
- Set up keep alive timeout
```

File: doc/source/serve/http-guide.md [hunk 10]
```diff
@@ -183,3 +184,16 @@ Client disconnecting
...
 ```
+
+
+(serve-http-guide-keep-alive-timeout)=
+## Set keep alive timeout
+
+Serve uses a Uvicorn HTTP server internally to serve HTTP requests. By default, Uvicorn
+keeps HTTP connections alive for 5 seconds between requests. Modify the keep-alive
+timeout by setting the `keep_alive_timeout_s` in the `http_options` field of the Serve
+config files. This config is global to your Ray cluster, and it cannot be updated during
```
Comment (angelinalg) [inline doc/source/serve/http-guide.md:195]:
```suggestion
config files. This config is global to your Ray cluster, and you can't update it during
```

File: doc/source/serve/http-guide.md [hunk 11]
```diff
@@ -183,3 +184,16 @@ Client disconnecting
...
+runtime. You can also set `RAY_SERVE_HTTP_KEEP_ALIVE_TIMEOUT_S` environment variable to
```
Comment (angelinalg) [inline doc/source/serve/http-guide.md:196]:
```suggestion
runtime. You can also set the `RAY_SERVE_HTTP_KEEP_ALIVE_TIMEOUT_S` environment variable to
```

File: doc/source/serve/http-guide.md [hunk 12]
```diff
@@ -183,3 +184,16 @@ Client disconnecting
...
+set the keep alive timeout. `RAY_SERVE_HTTP_KEEP_ALIVE_TIMEOUT_S` will take the
```
Comment (angelinalg) [inline doc/source/serve/http-guide.md:197]:
```suggestion
set the keep alive timeout. `RAY_SERVE_HTTP_KEEP_ALIVE_TIMEOUT_S` takes
```

File: doc/source/serve/http-guide.md [hunk 13]
```diff
@@ -183,3 +184,16 @@ Client disconnecting
...
+precedence over the `keep_alive_timeout_s` config if both are set. You can find more
```
Comment (angelinalg) [inline doc/source/serve/http-guide.md:198]:
```suggestion
precedence over the `keep_alive_timeout_s` config if both are set. See
```

File: doc/source/serve/http-guide.md [hunk 14]
```diff
@@ -183,3 +184,16 @@ Client disconnecting
...
+information about Uvicorn's keep alive timeout [here](https://www.uvicorn.org/settings/#keep-alive-timeout).
```
Comment (angelinalg) [inline doc/source/serve/http-guide.md:199]:
```suggestion
Uvicorn's keep alive timeout [guide](https://www.uvicorn.org/settings/#keep-alive-timeout) for more information.
```

Comment (angelinalg) [reply doc/source/serve/http-guide.md:199]: I think this reference is broken. Did you mean: https://www.uvicorn.org/settings/#timeouts ?

File: doc/source/serve/production-guide/config.md [hunk 15]
```diff
@@ -45,6 +47,7 @@ The file contains `http_options` and `applications`. These are the `http_options
 
 - `host` and `port` are HTTP options that determine the host IP address and the port for your Serve application's HTTP proxies. These are optional settings and can be omitted. By default, the `host` will be set to `0.0.0.0` to expose your deployments publicly, and the port will be set to `8000`. If you're using Kubernetes, setting `host` to `0.0.0.0` is necessary to expose your deployments outside the cluster.
```
Comment (angelinalg) [inline doc/source/serve/production-guide/config.md:48]:
```suggestion
- `host` and `port` are HTTP options that determine the host IP address and the port for your Serve application's HTTP proxies. These are optional settings and can be omitted. By default, the `host` is set to `0.0.0.0` to expose your deployments publicly, and the port is set to `8000`. If you're using Kubernetes, setting `host` to `0.0.0.0` is necessary to expose your deployments outside the cluster.
```

File: doc/source/serve/production-guide/config.md [hunk 16]
```diff
@@ -45,6 +47,7 @@ The file contains `http_options` and `applications`. These are the `http_options
...
 - `request_timeout_s` is a field in the `http_options` that allows you to set the end-to-end timeout for a request before terminating and retrying at another replica. This config is global to your Ray cluster, and it cannot be updated during runtime. By default, the Serve HTTP proxy retries up to `10` times when a response is not received due to failures (e.g. network disconnect, request timeout, etc.). By default, there is no request timeout.
```
Comment (angelinalg) [inline doc/source/serve/production-guide/config.md:49]:
```suggestion
- `request_timeout_s` is a field in the `http_options` that allows you to set the end-to-end timeout for a request before terminating and retrying at another replica. This config is global to your Ray cluster, and you can't update it during runtime. By default, the Serve HTTP proxy retries up to `10` times when a response is not received due to failures (for example, network disconnect, request timeout, etc.) By default, there is no request timeout. 
```

File: doc/source/serve/production-guide/config.md [hunk 17]
```diff
@@ -45,6 +47,7 @@ The file contains `http_options` and `applications`. These are the `http_options
 
 - `host` and `port` are HTTP options that determine the host IP address and the port for your Serve application's HTTP proxies. These are optional settings and can be omitted. By default, the `host` will be set to `0.0.0.0` to expose your deployments publicly, and the port will be set to `8000`. If you're using Kubernetes, setting `host` to `0.0.0.0` is necessary to expose your deployments outside the cluster.
 - `request_timeout_s` is a field in the `http_options` that allows you to set the end-to-end timeout for a request before terminating and retrying at another replica. This config is global to your Ray cluster, and it cannot be updated during runtime. By default, the Serve HTTP proxy retries up to `10` times when a response is not received due to failures (e.g. network disconnect, request timeout, etc.). By default, there is no request timeout. 
+- `keep_alive_timeout_s` is a field in the `http_options` that allows you to set the keep alive timeout for the HTTP proxy. This config is global to your Ray cluster, and it cannot be updated during runtime. By default, the Serve HTTP proxy keeps the connection alive for `5` seconds.
```
Comment (angelinalg) [inline doc/source/serve/production-guide/config.md:50]:
```suggestion
- `keep_alive_timeout_s` is a field in the `http_options` that allows you to set the keep alive timeout for the HTTP proxy. This config is global to your Ray cluster, and you can't update it during runtime. By default, the Serve HTTP proxy keeps the connection alive for `5` seconds.
```

**Review by angelinalg — APPROVED** [review]
Left some suggestions.

File: doc/source/serve/http-guide.md (hunk 14 above)
Comment (GeneDer) [reply doc/source/serve/http-guide.md:199]: oh interesting, I was meant to link to https://www.uvicorn.org/server-behavior/#timeouts but somehow got the wrong one. Thanks for catching it!

File: doc/source/serve/production-guide/config.md [hunk 18]
```diff
@@ -43,8 +45,9 @@ applications:
...
+- `host` and `port` are HTTP options that determine the host IP address and the port for your Serve application's HTTP proxies. These are optional settings and can be omitted. By default, the `host` is set to `0.0.0.0` to expose your deployments publicly, and the port is set to `8000`. If you're using Kubernetes, setting `host` to `0.0.0.0` is necessary to expose your deployments outside the cluster.
+- `request_timeout_s` is a field in the `http_options` that allows you to set the end-to-end timeout for a request before terminating and retrying at another replica. This config is global to your Ray cluster, and you can't update it during runtime. By default, the Serve HTTP proxy retries up to `10` times when a response is not received due to failures (for example, network disconnect, request timeout, etc.) By default, there is no request timeout. 
+- `keep_alive_timeout_s` is a field in the `http_options` that allows you to set the keep alive timeout for the HTTP proxy. This config is global to your Ray cluster, and you can't update it during runtime. By default, the Serve HTTP proxy keeps the connection alive for `5` seconds.
```
Comment (sihanwang41) [inline doc/source/serve/production-guide/config.md:50]:
```suggestion
- `keep_alive_timeout_s` is a field in the `http_options` that allows you to set the keep alive timeout for the HTTP proxy. For more details, see [here](serve-http-guide-keep-alive-timeout)
```

**Review by sihanwang41 — APPROVED** [review]

**Review by edoakes — APPROVED** [review]

**Review by akshay-anyscale — APPROVED** [review]

**Pull request merged**